### PR TITLE
refactor(layopt) & feature(classes)

### PR DESCRIPTION
**Pull request by ns-rse**
- Moves more code out of `layopt.trussopt()` into small functions that have unit tests.
- These functions, which are used in creating the structure, now reside in a new module `structure`. This is required because
otherwise we encounter circular arguments when moving to the `Structure` dataclass which uses the `__post_init__()`
method to generate the structure elements on instantiation as it would have required functions that resided in the
`layopt` module.
- Introduces the `Structure` dataclass which holds the structures/features, this has a minimal requirement of
  `parameters` argument which is itself the data class `Parameters` (loaded from the `default_config.yaml`). It has a
  `__post_init__()` method which does what its name suggests and includes calls to the series of commands that build the
  structure that now reside in the `structure` module, setting these as attributes to the `Structure` dataclass
- Includes tests for new classes and moves tests from `test_layopt.py` to `test_structure.py` where functions now reside
as well as a test of the `Structure` class (this uses mocking and patching to isolate its testing).

Currently `Structure` is _not_ used in the processing done by `trussopt()` introducing that will be a separate piece of
work.

@fisher568 I'd be particularly grateful of careful checking of the mocking and patching aspect of the `Structure` class,
its not something I've a great deal of experience or practice doing so a check that its meaningful set of tests would be
appreciated, thanks.

---

Before submitting a Pull Request please check the following.

- [X] Existing tests pass.
- [ ] ~Documentation has been updated and builds.~
- [X] Pre-commit checks pass.
- [X] New functions/methods have typehints and docstrings.
- [X] New functions/methods have tests which check the intended behaviour is correct.

### CODE REVIEW

File: src/layopt/structure.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    npt.NDArray[np.float64]
+        Two dimensional array of node coordinates.
+    """
+    xv, yv = np.meshgrid(range(width + 1), range(height + 1))
+    points = [Point(xv.flat[i], yv.flat[i]) for i in range(xv.size)]
+    logger.debug(f"Points created : {len(points)=}")
+    return np.array([[pt.x, pt.y] for pt in points if polygon.intersects(pt)])
+
+
+def calc_loaded_points(width: int, height: int) -> npt.NDArray[np.float64]:
```
Comment (fisher568) [inline src/layopt/structure.py]: I think should probably have `default` in the name and relevant bits of the docstring. We only use this if `loaded_points` is `None`, otherwise it's what is set in the config file.
On that note, perhaps we need to add a test for `loaded_points` not being set

File: tests/test_classes.py [hunk 2]
```diff
@@ -186,3 +189,136 @@ def test_parameters_exceptions(config: ValidationError) -> None:
...
+    # Get parameters fixture
+    parameters = request.getfixturevalue(fixture_str)
+
+    # Patch the make_polygon method using shapely.geometry.Polygon so that Pydantic validation passes
+    shapely_polygon = Polygon(bounding_coordinates)
+
+    def fake_make_polygon(bbox):  # noqa: ARG001, pylint: disable=unused-argument
+        return shapely_polygon
+
+    monkeypatch.setattr(structure_module, "make_polygon", fake_make_polygon)
```
Comment (fisher568) [inline tests/test_classes.py:279]: Would it not be easier to have this attribute set to `lambda: Polygon(bounding_coordinates)` and delete the `fake_make_polygon` function and the line above it? Cf what I've done in `test_reconciled_solver` in `test_config` with `mock_installed_solvers`

File: tests/test_classes.py [hunk 3]
```diff
@@ -186,3 +189,136 @@ def test_parameters_exceptions(config: ValidationError) -> None:
...
+                "all_patterns": np.asarray([[0, 0]]),
+                "potential_members": np.empty((0, 4)),
+                "primal_adaptivity": True,
+                "active_load_cases": np.asarray([0]),
+            },
+            id="18x4 spanning",
+        ),
+    ],
+)
+def test_structure(
```
Comment (fisher568) [inline tests/test_classes.py]: perhaps this name should be very slightly changed (to `test_make_structure` or something) since we have a file with that name and running `pytest -k test_structure` runs this test and everything in that file

File: tests/test_classes.py [hunk 4]
```diff
@@ -186,3 +189,136 @@ def test_parameters_exceptions(config: ValidationError) -> None:
...
+
+    # Patch the make_polygon method using shapely.geometry.Polygon so that Pydantic validation passes
+    shapely_polygon = Polygon(bounding_coordinates)
+
+    def fake_make_polygon(bbox):  # noqa: ARG001, pylint: disable=unused-argument
+        return shapely_polygon
+
+    monkeypatch.setattr(structure_module, "make_polygon", fake_make_polygon)
+
+    # Patch out the rest of __post_init__ derived attributes
```
Comment (fisher568) [inline tests/test_classes.py:281]: The patching/mocking looks good to me but I think maybe we should have them be more realistic values (and changed expected values accordingly of course). Don't have to be exact, but especially if errors creep in with the current sets of zeros everywhere it might not be the easiest to see what's going on

**Review by fisher568 — CHANGES_REQUESTED** [review]
Generally looks good, some inline comments

File: tests/test_classes.py (hunk 3 above)
Comment (ns-rse) [reply tests/test_classes.py]: Good suggestion, have gone with `test_structure_post_init(...)` which reflects the docstring and what we are actually testing.

File: src/layopt/structure.py (hunk 1 above)
Comment (ns-rse) [reply src/layopt/structure.py]: Gone with `calc_default_loaded_points()`.

I think adding the test for `loaded_points` being `None` should be a separate issue. Not sure how often that might arise but sensible to check it.

File: tests/test_classes.py (hunk 2 above)
Comment (ns-rse) [reply tests/test_classes.py:279]: Probably, I don't use `lambda` functions much really so didn't think to do so here.

I gave it a whirl though and switched to...

```suggestion
    monkeypatch.setattr(structure_module, "make_polygon", lambda: Polygon(bounding_coordinates))
```

On running the test it failed with...

```
>       self.polygon = structure.make_polygon(self.bounding_coordinates)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E       TypeError: test_structure_post_init.<locals>.<lambda>() takes 0 positional arguments but 1 was given
```

So I tried simply assigning first without putting it in a function...

```suggestion
    shapely_polygon = Polygon(bounding_coordinates)
    monkeypatch.setattr(structure_module, "make_polygon", shapely_polygon)
```

...which resulted in the same error.

Reverting to the original in this pull request and the tests pass.

Not sure why that fails here though but is acceptable in `tests_classes.py::test_reconcile_solver`, perhaps its down to patching with a list v a function. Often when I've used `lambda` functions its of the form `lambda x: x > 1.0`.

Something for me to go and learn more on and correct in the future I think.

File: tests/test_classes.py (hunk 4 above)
Comment (ns-rse) [reply tests/test_classes.py:281]: Happy to use something more realistic, and ideally minimal for the mocking, do you have any recommendations?

Comment (ns-rse) [reply tests/test_classes.py:281]: I think there are perhaps only really two parameters that would make much difference, `convex` and `primal_adaptivity` as these can both be either `True` / `False`. The other parameters that are mocked would, I think, just be longer arrays so changing those probably doesn't make much difference.

File: tests/test_classes.py (hunk 2 above)
Comment (fisher568) [reply tests/test_classes.py:279]: Strange, I'm not sure why that's happening either at first glance, will have a think but agree that original is fine for now at least

File: tests/test_classes.py (hunk 4 above)
Comment (fisher568) [reply tests/test_classes.py:281]: > Happy to use something more realistic, and ideally minimal for the mocking, do you have any recommendations?

Sorry, missed this reply at the time. Might have a think about this in future but it's OK for now. Changing one of the `convex` values to `False` and ditto for `primal_adaptivity` probs a good call though

**Review by fisher568 — APPROVED** [review]
Good to merge now. Sorry for whatever reason the emails didn't pop up as unread in my inbox which is why I missed this
